keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1077 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this enhancement, Kubernetes Developers will declare API validation using
+IDL tags in Go. For example:
+
+```go
+// staging/src/k8s.io/api/example/v1/types.go
+
+type FizzBuzzSpec struct {
+  // ...
+  //
+  // +maxLength=64
+  Name string `json...`
+  // ...
+}
+
+type FizzBuzz struct {
+  // ...
+  //
+  // +validations=rule:"self.replicas <= self.maxReplicas", messageExpression:"'replicas must be less than maxReplicas value of ' + string(self.maxReplicas)"
+  Spec FizzBuzzSpec `json...`
+}
+```
+
+In this example, both `+maxLength` and `+validations` are IDL tags.
+
+The declarative validation will be used by the kube-apiserver to
+validate API requests.
+
+The declarative validation will also be included in the published OpenAPI:
+
+```yaml
+openAPIV3Schema:
+    type: object
+    properties:
+      spec:
+        type: object
+        x-kubernetes-validations:
+          - rule: "self.replicas <= self.maxReplicas"
+       properties:
+         name:
+           type: string
+           maxLength: 64
+```
+
+It is important to note that the declarative validation will use the same set of
+validation options available today for CRDs. namely:
+
+- [JSON Schema value validations](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00) (e.g. `format:datetime`, `maxItems:64`)
```

Comment:
`//+maxItems=64` is Go comments, `maxItems: 64` in OpenAPI YAML.

I can show all the OpenAPI examples in JSON instead of YAML. That might minimize confusion between CRDs.